[CLIN-1078] HAPI FHIR Interceptor: Kafka Producer

czhu-carecloud opened the pull request.

apoggi-carecloud commented on src/main/resources/hapi.properties:
```diff
@@ -135,4 +135,13 @@ jwt.secret=secret123!
+##################################################
+
+kafka.host=pkc-epgnk.us-central1.gcp.confluent.cloud:9092
+kafka.key=NZ4XF4A7JAPN53PK
```
Let's move this to a .env file, .gitignore it, and commit the keys to a .env.sample file. I'd use this dependency for loading from the .env: https://github.com/cdimascio/java-dotenv. We don't want to commit these secrets to version control.

czhu-carecloud replied:
Is it ok if I just remove the values from the hapi.properties file? before committing? Is this issue similar to the `jwt.secret` in the same file?

apoggi-carecloud replied:
we should probably remove the jwt secret from the properties file also. We need to be able to set  and retrieve them per environment, so using dotenv is probably the best approach since its pretty much used across all of our projects

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+import ca.uhn.fhir.jpa.starter.context.ContextService;
+import ca.uhn.fhir.jpa.starter.context.models.Context;
+
+public class KafkaInterceptor {
```
A comment explaining the purpose of the interceptor would be useful for others

czhu-carecloud replied:
added

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+    RequestTypeEnum requestType = theRequestDetails.getRequestType();
+    String token = theRequestDetails.getHeader(CONTEXT);
+    ContextService contextService = new ContextService(cache, httpClient, token);
+    contextService.verify();
```
You're doing a context verification on every request here before the resource is returned. Context verification is already done by the ContextInterceptor on every resource save, so this is a bit unnecessary and is a performance hit.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+
+  FHIRKafkaProducer fhirKafkaProducer = new FHIRKafkaProducer();
+
+  private ILRUCache<String, List<String>> cache;
```
I don't see the purpose of having a cache or http client as fields of this class if they're never used.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+
+        String resourceId = theResource.getIdElement().toString().split("[/]")[1];
+
+        message.setType(requestType.toString());
```
Instead of using all these setters, you can add these fields to the constructor of a FHIRKafkaMessage and set them there. It's a bit more concise and moves more implementation details to the appropriate class.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  @Override
+  public void publish(String topic, String key, FHIRKafkaMessage message) {
```
you could log your publishing here

czhu-carecloud replied:
publishing is logged in the callback method of the `FHIRKafkaProducer`

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/IKafkaProducer.java:
```diff
@@ -0,0 +1,7 @@
+
+import ca.uhn.fhir.jpa.starter.kafka.models.FHIRKafkaMessage;
+
+public interface IKafkaProducer {
```
Makes more sense for this to be an IProducer, that way if we switch to another implementation like GCP Pub/Sub, the interface can be reused and it still makes sense.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+public class KafkaInterceptor {
+  private static final Logger log = LoggerFactory.getLogger(KafkaInterceptor.class.getName());
+
+  FHIRKafkaProducer fhirKafkaProducer = new FHIRKafkaProducer();
```
I would dependency inject an IProducer instead and assign it in the constructor. Look at how I injected the cache / http client in context interceptor if you need a reference. This makes for a more testable class that can have its components swapped out more easily.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+  private TestCallback callback;
+
+  public FHIRKafkaProducer() {
+    java.util.Map<String, String> env = System.getenv();
```
You're not using this Map on env vars anywhere. No point in calling this.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+  public FHIRKafkaProducer() {
+    java.util.Map<String, String> env = System.getenv();
+
+    String bootstrapServerConfig = HapiProperties.getKafkaHost();
```
Lets switch to dotenv package for these properties. Look at my comment in the .properties file for reference

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+    producerProps = new Properties();
+    producerProps.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, bootstrapServerConfig);
+    producerProps.put(ProducerConfig.ENABLE_IDEMPOTENCE_CONFIG, "true");
+    producerProps.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, "1");
```
are you sure you want 1 in flight req per connection? this could have performance implications.

apoggi-carecloud replied:
Please address

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+    producerProps.put(ProducerConfig.MAX_IN_FLIGHT_REQUESTS_PER_CONNECTION, "1");
+    producerProps.put(ProducerConfig.RETRIES_CONFIG, String.valueOf(Integer.MAX_VALUE));
+    producerProps.put(ProducerConfig.ACKS_CONFIG, "all");
+    producerProps.put(ProducerConfig.BATCH_SIZE_CONFIG, "0");
```
Not sure if this disables batching , but it would be good to have a batching strategy for the producer.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,70 @@
+        message.setMeta(resourceId);
+        message.setTimeStamp(timeStamp);
+
+        fhirKafkaProducer.publish(resourceType, resourceId, message);
```
Maybe consider leveraging the concurrency library in Java to do this async. https://docs.oracle.com/javase/tutorial/essential/concurrency/

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,76 @@
+  }
+
+  @Override
+  public void publish(String topic, String key, FHIRKafkaMessage message) {
```
writing a benchmark would be very useful for this method. We could run it and see how our configurations would affect the runtime of the method and possibly adjust the batching strategy.

apoggi-carecloud replied:
please address

czhu-carecloud commented:
- Implemented `java-dotenv` and moved kafka/jwt secrets to `.env` file.
- Removed assortment of unneeded implementations
- Created and implemented constructors for `kafka/models`
- Generalized IKafkaProducer to IProducer
- Dependency injected FHIRKafkaProducer into KafkaInterceptor
- TODO: Create benchmarking to help with optimizing Kafka producer props and concurrency implementations.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java:
```diff
@@ -497,6 +502,21 @@ public static String getAddressSettingsApi() {
     return HapiProperties.getProperty(ADDRESS_SETTINGS_API, null);
   }
 
+  public static String getKafkaHost() {
```
these getters are no longer necessary

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/HapiProperties.java:
```diff
@@ -76,6 +76,11 @@
   static final String JWT_SECRET = "jwt.secret";
   static final String ADDRESS_SETTINGS_API = "address.settings_api";
   static final String CACHE_MAX_CAPACITY = "cache.max_capacity";
+  static final String KAFKA_HOST = "kafka.host";
```
the kafka and jwt fields are no longer necessary

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/interceptors/KafkaInterceptor.java:
```diff
@@ -0,0 +1,62 @@
+
+  private IProducer producer;
+
+  public KafkaInterceptor(FHIRKafkaProducer producer){
```
Your constructor should take an IProducer. We don't want it to be implementation specific.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,75 @@
+  private final String bootstrapServerConfig = dotenv.get("KAFKA_HOST");
+  private final String KAFKA_KEY = dotenv.get("KAFKA_KEY");
+  private final String KAFKA_SECRET = dotenv.get("KAFKA_SECRET");
+  private static final String jaasTemplate = "org.apache.kafka.common.security.plain.PlainLoginModule required username=\"%s\" password=\"%s\";";
```
constants should be screaming snake case

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,75 @@
+  private static final Logger log = LoggerFactory.getLogger(FHIRKafkaProducer.class.getName());
+  private static final Dotenv dotenv = Dotenv.load();
+
+  private final String bootstrapServerConfig = dotenv.get("KAFKA_HOST");
```
I would initialize these values in the constructor

czhu-carecloud replied:
These configs are passed into the actual Producer extended from `org.apache.kafka.clients.producer.*` as `producerProps`. From the tutorial that I found, it's suggested to simply create a Properties hashmap in our Producer implementation's constructor to just pass to the `KafkaProducer`.

apoggi-carecloud commented on src/main/java/ca/uhn/fhir/jpa/starter/kafka/FHIRKafkaProducer.java:
```diff
@@ -0,0 +1,75 @@
+
+  @Override
+  public void publish(String topic, String key, FHIRKafkaMessage message) {
+    String serializedMessage = message.toString();
```
are you sure you want to use .toString() and not gson.toJson()?

czhu-carecloud replied:
I was having difficulty implementing the gson.toJson() so I created a custom deserializer or the FHIRKafkaMessage model
